SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialDispatcher.cpp

```diff
@@ -201,3 +172,108 @@ Worker_ComponentId USpatialDispatcher::GetComponentId(Worker_Op* Op) const
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_REMOVE_COMPONENT, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->remove_component);
+	});
+}
+
+USpatialDispatcher::CallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const UserCallback<Worker_AuthorityChangeOp>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_AUTHORITY_CHANGE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->authority_change);
+	});
+}
+USpatialDispatcher::CallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const UserCallback<Worker_ComponentUpdateOp>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMPONENT_UPDATE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->component_update);
+	});
+}
+
+USpatialDispatcher::CallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const UserCallback<Worker_CommandRequestOp>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_REQUEST, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->command_request);
+	});
+}
+
+USpatialDispatcher::CallbackId USpatialDispatcher::AddOpCallback(Worker_ComponentId ComponentId, const UserCallback<Worker_CommandResponseOp>& Callback)
+{
+	return AddGenericOpCallback(ComponentId, WORKER_OP_TYPE_COMMAND_RESPONSE, [Callback](const Worker_Op* Op)
+	{
+		Callback(Op->command_response);
+	});
+}
+
+
+USpatialDispatcher::CallbackId USpatialDispatcher::AddGenericOpCallback(Worker_ComponentId ComponentId, Worker_OpType OpType, const UserCallback<const Worker_Op*>& Callback)
+{
+	check(SpatialConstants::MIN_EXTERNAL_SCHEMA_ID <= ComponentId && ComponentId <= SpatialConstants::MAX_EXTERNAL_SCHEMA_ID);
+	const CallbackId NewCallbackId = NextCallbackId++;
+	ComponentOpTypeToCallbackIdMap.FindOrAdd(ComponentId).FindOrAdd(OpType).Add(UserOpCallbackData{ NewCallbackId, Callback });
+	CallbackIdToDataMap.Add(NewCallbackId, CallbackIdData{ ComponentId, OpType });
+	return NewCallbackId;
+}
+
+void USpatialDispatcher::RemoveOpCallback(CallbackId CallbackId)
+{
+	// Find callback ID in map and assert if it does not exist
+	CallbackIdData CallbackDataToRemove = CallbackIdToDataMap.FindAndRemoveChecked(CallbackId);
+
+	TMap<Worker_OpType, TArray<UserOpCallbackData>>* OpTypesToCallbackIds = ComponentOpTypeToCallbackIdMap.Find(CallbackDataToRemove.ComponentId);
+	check(OpTypesToCallbackIds);
+	TArray<UserOpCallbackData>* ComponentCallbackIds = OpTypesToCallbackIds->Find(CallbackDataToRemove.OpType);
+	check(ComponentCallbackIds);
+
+	// If removing only callback for a component ID / op type, delete map entries as applicable
+	if (OpTypesToCallbackIds->Num() == 1)
+	{
+		if (ComponentOpTypeToCallbackIdMap.Num() == 1)
+		{
+			ComponentOpTypeToCallbackIdMap.Remove(CallbackDataToRemove.ComponentId);
+			return;
+		}
+		OpTypesToCallbackIds->Remove(CallbackDataToRemove.OpType);
+		return;
+	}
+
+	ComponentCallbackIds->RemoveAll([CallbackId](const UserOpCallbackData& Data)
```

Comment:
As we know we can only find it once it might be easier to just do a for loop.

Reply:
Why easier?

Reply:
All I just use RemoveAll because it's the only method I could find that takes a predicate as the argument

Reply:
Easier was the wrong word. I meant I personally would find it easier to read as it's more commonly seen code. Doesn't create more objects etc. I think there is also a RemoveSingle that would just take a uint.

Reply:
but it's a struct containing int + Tfunc? happy to migrate to loop either way if you think it's worth changing?

Reply:
Oh yes, we stopped it being just an int. In which case I'd do a loop.